### ISSUE TITLE
Should ActiveRecord::Relation#null_relation? be made public?

### ISSUE THREAD

**Issue by ziggythehamster** [rails/rails#45291]
I will admit this is a bit of a niche use-case, but there is currently (to my knowledge, at least) no _public_ way to determine if a relation is `none?` without executing it in the case that it is not a null relation. The protected method `null_relation?` does the right thing, but it's protected, so one must use a refinement to change its visibility. You could alternatively do `.is_a?(ActiveRecord::NullRelation)` but that seems worse.

My specific use case is that I have a scope which will short-circuit by returning `none` when the argument to that scope doesn't exist in a look-up table. I don't want to expose the details of this look-up table to callers because there are multiple callers and I don't want to repeat myself.

This scope is used in a subquery, and if the scope is `none?` we can avoid running that outer query as well. The outer query is pretty complex and involves another few queries and some joins, and it will return zero rows if the inner query is `none?`. There are also a few of these outer queries.

What we landed on is to create a refinement that implements the existing `null_relation?` method:

```ruby
      def null_relation?
        is_a?(ActiveRecord::NullRelation)
      end
```

...and then we short-circuit the outer query by going `return none if subquery_scope.null_relation?`. This works exactly the same as `return none if subquery_scope.none?`, except that it doesn't execute the query when it is something.

The simplest way to make this advanced usage work is to make the protected method public, but this might confuse new users who don't know what the difference is between `none?` and `null_relation?`. I don't like the idea of doing the `is_a?` check every time (because then `NullRelation` leaks into our code in more than one place). I also think that doing something like `none?(execute: false)` is potentially surprising (it returns false when it may actually still be false if you called `.none?`), but this might be a good enough compromise in the API if others feel that `null_relation?` is potentially confusing to new users.

I did some searching and I didn't see any discussion on whether this should be public or not in the context of more modern versions of Rails. It seems like when this was introduced, it was a performance optimization primarily, and that's why the method was made protected, but that's about all I came across. I'm curious what others' thoughts are about this today.

**Comment by ziggythehamster** [comment]
I can try to do that. The essential issue is that `none?` will execute the query (with an `EXISTS` in Postgres) if a scope is _not_ a `NullRelation`, and will not execute when it is. If you have a query that uses a scope in a subquery, where that scope can simply `return none` your options are:

* Execute the outer and inner queries in all cases, knowing the inner query could return no results without ever needing to execute, and therefore the outer query returns no results, but it had to execute a query to find that out. This has one query always.
* Execute the inner query once if it isn't a `NullRelation` to determine if you can `return none if subquery.none?`, and then execute it again when it's in the subquery. This has zero queries if it's a `NullRelation`, one query if the results are empty but it's not a `NullRelation`, and two if it has results and isn't a `NullRelation`.

`.none?(execute: false)` would be sufficient to make the second case be zero or one query (but the same applies to `.null_relation?`). I agree that exposing `.null_relation?` has a smell, so I would prefer something different. I'm just not entirely sure what that is.

It also bears reiterating that like `.optimizer_hints`, this is an advanced use of ActiveRecord. In general, you would want to execute multiple queries because it's easier to reason about. In our case, we know whether zero rows or possibly some rows will be returned without doing a query, and if zero rows are expected, we don't need to run the outer query. `.none?` works for this only when there are zero rows, otherwise it executes a query to determine if it's none and then executes another query containing this query if it isn't none.

Basically, `none?` is a boolean that handles three states - null relation, no results, some results. But we only need to know if it's a null relation or not.

**Comment by ziggythehamster** [comment]
> Would it be possible for us to internally recognise that the inner query being a NullRelation renders the outer's condition impossible?

Possibly, but I'm not entirely sure what that would look like. I don't know if I would always want a filter by `relation.none` to kill the query.

> Is there also some surrounding context as to why one vs zero queries is a worthwhile saving? In most application circumstances, given the background overhead of authorization, I wouldn't really expect a single query saving to be worth bothering with the conditional.

These are really huge query results, and decomposing them in ActiveRecord means we have to marshal tens of thousands of rows which Postgres only needs to find a few of if we weren't doing that. So, instead, the query is complex but allows PG to leverage its indexes to turn the huge possible set into only a few.

Here's our use case, generalized. Our query is way more complicated, and we have split it up a bit differently, so I may have made a mistake bringing it over.

```ruby
# This is on Foo
def outer_scope(lookup_key, a_ids)
    return none if inner_scope(lookup_key).null_relation? # this is using our refinement that makes the method public

    optimizer_hints('HashJoin(foo bar)')
            .select(:one, :two)
            .from(
              Bar.where(a_id: a_ids, a_type: 'X', b_type: 'Y') # This is a HABTM join table, essentially
                .select('bar.a_id, foo.*, ROW_NUMBER() OVER (PARTITION BY bar.a_id ORDER BY foo.x DESC)') # we build this with arel but this is simplified
                .joins('foo ON foo.a_id = bar.a_id') # same here, so the syntax might be wrong here
                .merge(inner_scope(lookup_key))
            )
            .where(row_number: 1)
end

# This is the inner scope (also on Foo). Based on a static lookup, we can just return none. The actual code that does
# this will memoize the result of this method so we aren't building the Arel stuff twice
def inner_scope(lookup_key)
  things = SomeLookup.lookup(lookup_key) # this could be Redis, ENV, whatever
  return none if things.empty?

  where(
    Arel::Nodes::Grouping.new(
      Foo.arel_table[:one].eq('first').and(Foo.arel_table[:two].in(things[:first]))
    ).or(Arel::Nodes::Grouping.new(
      Foo.arel_table[:one].eq('second').and(Foo.arel_table[:two].in(things[:second]))
    ))
  )
end
```

In reality, there's a lot more composition going on, so while it seems like the obvious solution is to just do the lookup in the outer scope and do where in the `merge()`, this ends up not being possible. Also, these are implemented as query classes that wrap all of the complex windowing and filtering and joining in one place. The inner scope gets reused elsewhere (there are multiple queries that are concerned about filtering Postgres given the result of the lookup), and preferably we don't expose that implementation detail to things that want to use that scope.

Maybe what should happen is that if you attempt to `merge` a `NullRelation` it knows this will result in no results and makes the thing it's merged into a `NullRelation`? Which then means you're doing `.from(NullRelation)` and that's also a no-op.

**Comment by ziggythehamster** [comment]
Hm, and doing some more digging (Rails 6.1), I think that I'm needing `null_relation?` because there is a bug:

```
3.0.0 :001 > Foo.joins(:bars).merge(Bar.none).to_sql
 => ""
3.0.0 :002 > Foo.from(Foo.joins(:bars).merge(Bar.none)).to_sql
 => "SELECT \"foos\".* FROM (SELECT \"foos\".* FROM \"foos\" INNER JOIN \"bars\" ON \"bars\".\"id\" = \"foos\".\"bar_id\" WHERE (1=0)) subquery" 
```

Obviously the above simplification is a really silly way of doing that query, but it's the simplest version that exhibits the problem. Also, the SQL is incorrect; there needs to be an `.as`, but that doesn't actually matter. I know that null relations end up being `WHERE (1=0)` when forced to get themselves involved in execution, so most likely there's something that needs to check for the null relation in `from` that isn't.

**Comment by p8** [comment]
This seems to have been fixed by: https://github.com/rails/rails/pull/47920
